src/typegraph/graph/src/path.rs

```diff
@@ -0,0 +1,521 @@
+
+        left.base.type_idx == right.base.type_idx && self.path == other.path
+    }
+}
+
+impl Eq for ValueTypePath {}
+
+impl Hash for ValueTypePath {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.owner
+            .upgrade()
+            .expect("no strong pointer for type")
+            .base
+            .type_idx
+            .hash(state);
+        self.path.hash(state);
+    }
+}
+
+/// type id in the expanded graph
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum RelativePath {
+    Function(u32),
+    NsObject(Vec<Arc<str>>),
+    Input(ValueTypePath),
+    Output(ValueTypePath),
+}
+
+impl TryFrom<RelativePath> for ValueTypePath {
+    type Error = RelativePath;
+    fn try_from(value: RelativePath) -> Result<Self, Self::Error> {
+        match value {
+            RelativePath::Input(path) => Ok(path),
+            RelativePath::Output(path) => Ok(path),
+            _ => Err(value),
+        }
+    }
+}
+
+impl std::fmt::Debug for RelativePath {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Function(idx) => write!(f, "Function({})", idx),
+            Self::NsObject(path) => write!(f, "NsObject(/{:?})", path.join("/")),
+            Self::Input(k) => write!(
+                f,
+                "Input(fn={}; {:?})",
+                k.owner.upgrade().expect("no strong pointer for type").idx(),
+                k.path
+            ),
+            Self::Output(k) => write!(
+                f,
+                "Output(fn={}; {:?})",
+                k.owner.upgrade().expect("no strong pointer for type").idx(),
+                k.path
+            ),
+        }
+    }
+}
+
+impl RelativePath {
+    pub fn root() -> Self {
+        Self::NsObject(vec![])
+    }
+
+    pub fn namespace(path: Vec<Arc<str>>) -> Self {
+        Self::NsObject(path)
+    }
+
+    pub fn input(owner: Weak<FunctionType>, path: Path) -> Self {
+        Self::Input(ValueTypePath {
+            owner,
+            path,
+            branch: ValueTypeKind::Input,
+        })
+    }
+
+    pub fn output(owner: Weak<FunctionType>, path: Path) -> Self {
+        Self::Output(ValueTypePath {
+            owner,
+            path,
+            branch: ValueTypeKind::Output,
+        })
+    }
+
+    pub fn is_input(&self) -> bool {
+        matches!(self, Self::Input(_))
+    }
+
+    pub fn is_output(&self) -> bool {
+        matches!(self, Self::Output(_))
+    }
+
+    pub fn is_namespace(&self) -> bool {
+        matches!(self, Self::NsObject(_))
+    }
+
+    pub fn contains(&self, nodes: &[tg_schema::TypeNode], xkey: &TypeKeyEx) -> bool {
+        use RelativePath as RP;
+
+        match self {
+            RP::Function(_) => false,
+            RP::NsObject(_) => false,
+            RP::Input(p) => {
+                let owner = p.owner.upgrade().expect("no strong pointer for type");
+                let input_type = match &nodes[owner.base.type_idx as usize] {
```

Comment:
Does owner always outlive or is of the same lifetime as self?

Reply:
Ultimately, they all have at least a strong reference in the `Typegraph` object.
Therefore, while the typegraph lives, all the types live...